src/main/java/htsjdk/samtools/SAMUtils.java

```diff
@@ -111,57 +111,61 @@
     public static final int MAX_PHRED_SCORE = 93;
 
     /**
-     * Convert from a byte array containing =AaCcGgTtNn represented as ASCII, to a byte array half as long,
-     * with =, A, C, G, T converted to 0, 1, 2, 4, 8, 15.
+     * Convert from a byte array containing =AaCcGgTtNnMmRrSsVvWwYyHhKkDdBb represented as ASCII, to a byte array half as long,
+     * with fore example, =, A, C, G, T converted to 0, 1, 2, 4, 8, 15.
      *
      * @param readBases Bases as ASCII bytes.
+     * @param readName Name of the read containing the bases,
      * @return New byte array with bases represented as nybbles, in BAM binary format.
      */
-    static byte[] bytesToCompressedBases(final byte[] readBases) {
+    static byte[] bytesToCompressedBases(final byte[] readBases, final String readName) {
         final byte[] compressedBases = new byte[(readBases.length + 1) / 2];
         int i;
         for (i = 1; i < readBases.length; i += 2) {
-            compressedBases[i / 2] = (byte) (charToCompressedBaseHigh(readBases[i - 1]) |
-                    charToCompressedBaseLow(readBases[i]));
+            compressedBases[i / 2] = (byte) (charToCompressedBaseHigh(readBases[i - 1], readName) |
+                    charToCompressedBaseLow(readBases[i], readName));
         }
         // Last nybble
         if (i == readBases.length) {
-            compressedBases[i / 2] = charToCompressedBaseHigh((char) readBases[i - 1]);
+            compressedBases[i / 2] = charToCompressedBaseHigh(readBases[i - 1], readName);
         }
         return compressedBases;
     }
 
     /**
-     * Convert from a byte array with basese stored in nybbles, with =, A, C, G, T represented as 0, 1, 2, 4, 8, 15,
+     * Convert from a byte array with bases stored in nybbles, with fore example,=, A, C, G, T, N represented as 0, 1, 2, 4, 8, 15,
      * to a a byte array containing =AaCcGgTtNn represented as ASCII.
      *
      * @param length Number of bases (not bytes) to convert.
      * @param compressedBases Bases represented as nybbles, in BAM binary format.
+     * @param readName Name of the read containing the bases.
      * @param compressedOffset Byte offset in compressedBases to start.
      * @return New byte array with bases as ASCII bytes.
      */
-    public static byte[] compressedBasesToBytes(final int length, final byte[] compressedBases, final int compressedOffset) {
+    public static byte[] compressedBasesToBytes(final int length, final byte[] compressedBases, final int compressedOffset, final String readName) {
```

Comment:
I don't think it 's a great idea to change the signature of this public method to require the read name to produce a better error message in a specific use case.  I think it would be better to catch the exception in the calling method and re-throw an exception with the added information.